library/pkcs7.c

```diff
@@ -0,0 +1,525 @@
+            &signers_set->sig_alg_identifier );
+    if( ret != 0 )
+        return( ret );
+
+    ret = pkcs7_get_signature( p, end, &signers_set->sig );
+    if( ret != 0 )
+        return( ret );
+
+    signers_set->next = NULL;
+
+    return( 0 );
+}
+
+/**
+ * SignedData ::= SEQUENCE {
+ *      version Version,
+ *      digestAlgorithms DigestAlgorithmIdentifiers,
+ *      contentInfo ContentInfo,
+ *      certificates
+ *              [0] IMPLICIT ExtendedCertificatesAndCertificates
+ *                  OPTIONAL,
+ *      crls
+ *              [0] IMPLICIT CertificateRevocationLists OPTIONAL,
+ *      signerInfos SignerInfos }
+ */
+static int pkcs7_get_signed_data( unsigned char *buf, size_t buflen,
+        mbedtls_pkcs7_signed_data *signed_data )
+{
+    unsigned char *p = buf;
+    unsigned char *end = buf + buflen;
+    size_t len = 0;
+    int ret;
+    mbedtls_md_type_t md_alg;
+
+    ret = mbedtls_asn1_get_tag( &p, end, &len, MBEDTLS_ASN1_CONSTRUCTED
+            | MBEDTLS_ASN1_SEQUENCE );
+    if( ret != 0 )
+        return( MBEDTLS_ERR_PKCS7_INVALID_FORMAT + ret );
+
+    /* Get version of signed data */
+    ret = pkcs7_get_version( &p, end, &signed_data->version );
+    if( ret != 0 )
+        return( ret );
+
+    /* If version != 1, return invalid version */
+    if( signed_data->version != MBEDTLS_PKCS7_SUPPORTED_VERSION ) {
+        return( MBEDTLS_ERR_PKCS7_INVALID_VERSION );
+    }
+
+    /* Get digest algorithm */
+    ret = pkcs7_get_digest_algorithm_set( &p, end,
+            &signed_data->digest_alg_identifiers );
+    if( ret != 0 ) {
+        return( ret );
+    }
+
+    ret = mbedtls_oid_get_md_alg( &signed_data->digest_alg_identifiers, &md_alg );
+    if( ret != 0 )
+        return( MBEDTLS_ERR_PKCS7_INVALID_ALG + ret );
+
+    /* Do not expect any content */
+    ret = pkcs7_get_content_info_type( &p, end, &signed_data->content.oid );
+    if( ret != 0 )
+        return( ret );
+
+    if( MBEDTLS_OID_CMP( MBEDTLS_OID_PKCS7_DATA, &signed_data->content.oid ) ) {
+        return( MBEDTLS_ERR_PKCS7_INVALID_CONTENT_INFO ) ;
+    }
+
+    p = p + signed_data->content.oid.len;
+
+    /* Look for certificates, there may or may not be any */
+    ret = pkcs7_get_next_content_len( &p, end, &len );
+    if( ret == 0 ) {
+        mbedtls_x509_crt_init( &signed_data->certs );
+        ret = pkcs7_get_certificates( &p, len, &signed_data->certs );
+        if( ret != 0 )
+            return( ret ) ;
+
+      p = p + len;
+    }
+
+    /* TODO: optional CRLs go here */
+
+    /* Get signers info */
+    ret = pkcs7_get_signers_info_set( &p, end, &signed_data->signers );
+    if( ret != 0 )
+        return( ret );
+
+    return( ret );
+}
+
+int mbedtls_pkcs7_parse_der( const unsigned char *buf, const int buflen,
+        mbedtls_pkcs7 *pkcs7 )
+{
+    unsigned char *start;
+    unsigned char *end;
+    size_t len = 0;
+    int ret;
+
+    /* use internal buffer for parsing */
+    start = ( unsigned char * )buf;
```

Comment:
I know our style convention is a bit funny, but for casts we don't have spaces before/after brackets: `(unsigned char*) buf;`.